lib/compiler/src/beam_ssa_pre_codegen.erl

```diff
@@ -2064,32 +2064,63 @@ copy_retval_is([], RC, _, Copy, Count, Acc) ->
 %%
 %% Consider this code:
 %%
-%%   Var = ...
+%%   P = ...
+%%   Q = ...
 %%   ...
-%%   A1 = call foo/0
-%%   A = copy A1
-%%   B = call bar/1, Var
+%%   A = call foo/0
+%%   A1 = copy A
+%%   B = call bar/2, P, Q
 %%
-%% If the Var variable is no longer used after this code, its Y register
-%% can't be reused for A. To allow the Y register to be reused
-%% we will need to insert 'copy' instructions for arguments that are
-%% in Y registers:
+%% If the P or Q variables are no longer used after this code, one of
+%% their Y registers can't be reused for A. To allow one of the Y registers to
+%% be reused we will need to insert 'copy' instructions for arguments
+%% that are in Y registers:
 %%
-%%   Var = ...
+%%   P = ...
+%%   Q = ...
 %%   ...
 %%   A1 = call foo/0
-%%   Var1 = copy Var
+%%   Q1 = copy Q
+%%   P1 = copy P
+%%   A = copy A1
+%%   B = call bar/2, P1, Q1
+%%
+%% Note that copies of the arguments are done in reverse order to help the
+%% reserve_xregs/3 function place the copies into the X registers they will
+%% need to be in.
+%%
+%% For this example, P1 needs to be in x0 and Q1 needs to be in x1. If we
+%% would copy the arguments in order the registers would be assigned like
+%% this:
+%%
+%%   x0/A1 = call foo/0
+%%   x1/P1 = copy P
+%%   x2/Q1 = copy Q
 %%   A = copy A1
-%%   B = call bar/1, Var1
+%%   B = call bar/2, P1, Q1
+%%
+%% That is, both P1 and Q1 would be misplaced and would have to be
+%% moved to their correct registers before the call. However, with the
+%% copies in reverse order and with a little help from
+%% reserve_xregs/3, at least the Q1 variable can be can be placed in
+%% the correct register:
+%%
+%%   x0/A1 = call foo/0
+%%   x1/Q1 = copy Q
+%%   x2/P1 = copy P
+%%   A = copy A1
+%%   B = call bar/2, P1, Q1
+%%
+%% In general, all but the first argument can be placed in their correct registers.
 %%
 
 place_retval_copy(I, _Yregs, none, Count, Acc) ->
     {I,Count,Acc};
 place_retval_copy(#b_set{args=[F|Args0]}=I, Yregs, Copy, Count0, Acc0) ->
     #b_set{dst=Avoid} = Copy,
-    {Args,Acc1,Count} = copy_func_args(Args0, Yregs, Avoid, Acc0, [], Count0),
+    {Args,Acc1,Count} = copy_func_args(reverse(Args0), Yregs, Avoid, Acc0, [], Count0),
     Acc = [Copy|Acc1],
-    {I#b_set{args=[F|Args]},Count,Acc}.
+    {I#b_set{args=[F|reverse(Args)]},Count,Acc}.
```

Comment:
The base case of `copy_func_args/6` returns `{reverse(Acc),CopyAcc,Count}` and it's only called from here, so why not remove both reversals?

Reply:
That will not work. (Hint: The `reverse/1` at line 2123 undoes the `reverse/1` at line 2121.) One day I might clean up those functions and their creepy coupling.

Reply:
Sorry about the confusion, I meant the reversals at lines 2123 and 2137 (base case), not 2121 and 2123.

Reply:
You are right, that will work. I have done that simplification, along with another simplification and renaming of variables to make it clearer what is happening.